File: VIQA/HtmlElements/SimpleElements/Table.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using OpenQA.Selenium;
using VIQA.Common;
using VIQA.Common.Pairs;
using VIQA.HtmlElements.Interfaces;
using VIQA.SiteClasses;

namespace VIQA.HtmlElements.SimpleElements
{
    public class Table : Table<TextElement>, ITable { }

    public enum TableElementIndexType { Nums, Names, RowNames, ColNames }
    public enum TableHeadingType { ColumnsOnly, RowsOnly, NoHeadings, RowsAndColumns }

    public class Table<T> : VIElement, ITable<T> where T : VIElementsSet, IHaveValue
    {
        protected override string _typeName { get { return "Table"; } }

        private Dictionary<string, Dictionary<string, Cell<T>>> _allCells;
        public Dictionary<string, Dictionary<string, Cell<T>>> Cells
        {
            get
            {
                ColumnNames.ToDictionary(columnName => columnName, columnName =>
                        RowNames.ToDictionary(rowName => rowName, rowName =>
                            Cell(columnName, rowName)));
                return _allCells;
            }
        }

        public List<Cell<T>> GetColumn(int colIndex)
        {
            var result = new List<Cell<T>>();
            for (int rowIndex = 1; rowIndex <= RowCount; rowIndex++)
                result.Add(Cell(colIndex, rowIndex));
            return result.ToList();
        }

        public List<Cell<T>> GetColumn(string colName)
        {
            return RowNames.Select(rowName => Cell(colName, rowName)).ToList();
        }

        public List<Cell<T>> GetRow(int rowIndex)
        {
            var result = new List<Cell<T>>();
            for (int colIndex = 1; colIndex <= ColCount; colIndex++)
                result.Add(Cell(colIndex, rowIndex));
            return result.ToList();
        }
        public List<Cell<T>> GetRow(string rowName)
        {
            return ColumnNames.Select(colName => Cell(colName, rowName)).ToList();
        }

		private Func<IWebElement, string[]> _getColumnNamesFunc;
        public Func<IWebElement, string[]> GetColumnNamesFunc
        {
            set { _getColumnNamesFunc = value; }
            get { return _getColumnNamesFunc ??
                (table => table.FindElements(By.XPath(".//th")).Select(el => el.Text).ToArray());
            }
        }

        private int? _colCount;
        public int ColCount
        {
            set { _colCount = value; }
            get { return _colCount ?? (int)(_colCount = ColumnNames.Length); }
        }

        private string[] _columnNames;
        public string[] ColumnNames
        {
            set { _columnNames = value; }
            get
            {
                if (_columnNames != null)
                    return _columnNames;
                _columnNames = DoVIAction("GetColumnNames", () => GetColumnNamesFunc(GetWebElement()));
                if (_columnNames == null || !_columnNames.Any())
                    throw VISite.Alerting.ThrowError("Table have 0 columns. Please Specify ColumnNames or GetColumnNamesFunc");
                ColCount = _columnNames.Length;
                if (!HaveColumnNames)
                    _columnNames = GetNumList(ColCount);
                return _columnNames;
            }
        }
        
        private bool HaveColumnNames { get { return new[] { TableHeadingType.ColumnsOnly, TableHeadingType.RowsAndColumns }.Contains(HeadingsType); } }
        private bool HaveRowNames { get { return new[] { TableHeadingType.RowsOnly, TableHeadingType.RowsAndColumns }.Contains(HeadingsType); } }

        private Func<IWebElement, string[]> _getRowNamesFunc;
        public Func<IWebElement, string[]> GetRowNamesFunc
        {
            set { _getRowNamesFunc = value; }
            get
            {
                return _getRowNamesFunc ??
                    (table => table.FindElements(By.XPath(".//tr/td[1]")).Select(el => el.Text).ToArray());
            }
        }

        private int? _rowCount;
        public int RowCount { 
            set { _rowCount = value; }
            get { return _rowCount ?? (int)(_rowCount = RowNames.Length); }
        }

        private string[] _rowNames;
        public string[] RowNames
        {
            set { _rowNames = value; }
            get
            {
                if (_rowNames != null)
                    return _rowNames;
                _rowNames = DoVIAction("GetRowNames", () => GetRowNamesFunc(GetWebElement()));
                if (_rowNames == null || !_rowNames.Any())
                    throw VISite.Alerting.ThrowError("Table have 0 rows. Please Specify RowNames or GetRowNamesFunc");
                RowCount = _rowNames.Length;
                if (!HaveRowNames)
                    _rowNames = GetNumList(RowCount);
                return _rowNames;
            }
        }
        
        public int StartColumnIndex = 1;
        public int StartRowIndex = 1;

        public TableElementIndexType IndexType = TableElementIndexType.Nums;
        public TableHeadingType HeadingsType = TableHeadingType.ColumnsOnly;
        
        public Func<T> CellTemplate = () => (T)Activator.CreateInstance(typeof(T));
        private By _cellLocatorTemplate;

        private T CreateCell()
        {
            var cell = CellTemplate();
            cell.Context = new Pairs<ContextType, By>(ContextType.Locator, Locator, Context);
            return cell;
        }
        
        public Cell<T> Cell(int colNum, int rowNum)
        {
            var colIndex = colNum + StartColumnIndex - 1;
            var rowIndex = rowNum + StartRowIndex - 1;
            return CreateCell(colIndex, rowIndex, colNum, rowNum);
        }

        public Cell<T> Cell(string colName, string rowName)
        {
            var colIndex = GetColumnIndex(colName);
            var rowIndex = GetRowIndex(rowName);
            if (_allCells == null)
                _allCells = new Dictionary<string, Dictionary<string, Cell<T>>> { { colName, new Dictionary<string, Cell<T>> { { rowName, CreateCell(colIndex, rowIndex, colName, rowName) } } } };
            else 
                if (!_allCells.ContainsKey(colName))
                    _allCells.Add(colName, new Dictionary<string, Cell<T>> { { rowName, CreateCell(colIndex, rowIndex, colName, rowName) } });
                else if (!_allCells[colName].ContainsKey(rowName))
                    _allCells[colName].Add(rowName, CreateCell(colIndex, rowIndex, colName, rowName));
            return _allCells[colName][rowName];
        }
        
        public List<Cell<T>> FindCellsWithValue(string value)
        {
            return FindCellsWithValue(new Regex("^" + value + "$"));
        }

        public List<Cell<T>> FindCellsWithValue(Regex regex)
        {
            return Cells.SelectMany(col => col.Value.Where(row => regex.IsMatch(row.Value.Value)).Select(row => row.Value)).ToList();
        }

        private  bool GetCellFromValue(int colIndex, int rowIndex, string value, out Cell<T> cell)
        {
            cell = Cell(colIndex, rowIndex);
            return cell.Value == value;
        }

        public Cell<T> FindFirstCellWithValue(string value)
        {
            Cell<T> cell;
            for (int colIndex = 1; colIndex <= ColCount; colIndex++)
                for (int rowIndex = 1; rowIndex <= RowCount; rowIndex++)
                    if (GetCellFromValue(colIndex, rowIndex, value, out cell))
                        return cell;
            return null;
        }

        public Cell<T> FindCellInColumn(int colIndex, string value)
        {
            Cell<T> cell;
            for (int rowIndex = 1; rowIndex <= RowCount; rowIndex++)
                if (GetCellFromValue(colIndex, rowIndex, value, out cell))
                    return cell;
            return null;
        }

        public List<Cell<T>> FindCellsInColumn(int colIndex, Regex regex)
        {
            return GetColumn(colIndex).Where(cell => regex.IsMatch(cell.Value)).ToList();
        }

        public Cell<T> FindCellInColumn(string colName, string value)
        {
            Cell<T> cell;
            var colIndex = Array.IndexOf(ColumnNames, colName) + 1;
            for (int rowIndex = 1; rowIndex <= RowCount; rowIndex++)
                if (GetCellFromValue(colIndex, rowIndex, value, out cell))
                    return cell;
            return null;
        }

        public List<Cell<T>> FindCellsInColumn(string colname, Regex regex)
        {
            return GetColumn(colname).Where(cell => regex.IsMatch(cell.Value)).ToList();
        }

        public Cell<T> FindCellInRow(int rowIndex, string value)
        {
            Cell<T> cell;
            for (int colIndex = 1; colIndex <= ColCount; colIndex++)
                if (GetCellFromValue(colIndex, rowIndex, value, out cell))
                    return cell;
            return null;
        }

        public List<Cell<T>> FindCellsInRow(int rowIndex, Regex regex)
        {
            return GetRow(rowIndex).Where(cell => regex.IsMatch(cell.Value)).ToList();
        }

        public Cell<T> FindCellInRow(string rowName, string value)
        {
            Cell<T> cell;
            var rowIndex = Array.IndexOf(RowNames, rowName) + 1;
            for (int colIndex = 1; colIndex <= ColCount; colIndex++)
                if (GetCellFromValue(colIndex, rowIndex, value, out cell))
                    return cell;
            return null;
        }

        public List<Cell<T>> FindCellsInRow(string rowName, Regex regex)
        {
            return GetRow(rowName).Where(cell => regex.IsMatch(cell.Value)).ToList();
        }

        public List<Cell<T>> FindColumnByRowValue(int rowIndex, string value)
        {
            var columnCell = FindCellInRow(rowIndex, value);
            return columnCell != null ? GetColumn(columnCell.X) : null;
        }

        public List<Cell<T>> FindColumnByRowValue(string rowName, string value)
        {
            var columnCell = FindCellInRow(rowName, value);
            return columnCell != null ? GetColumn(columnCell.X) : null;
        }

        public List<Cell<T>> FindRowByColumnValue(int colIndex, string value)
        {
            var rowCell = FindCellInColumn(colIndex, value);
            return rowCell != null ? GetRow(rowCell.Y) : null;
        }

        public List<Cell<T>> FindRowByColumnValue(string colName, string value)
        {
            var rowCell = FindCellInColumn(colName, value);
            return rowCell != null ? GetRow(rowCell.Y) : null;
        }

        private int GetColumnIndex(string name)
        {
            int nameIndex;
            if (ColumnNames != null && ColumnNames.Contains(name))
                nameIndex = Array.IndexOf(ColumnNames, name);
            else
                throw VISite.Alerting.ThrowError("Can't Get Column: '" + name + "'. " + ((ColumnNames == null) ? "ColumnNames is Null" : ("Available ColumnNames: " + ColumnNames.Print(format: "'{0}'") + ")")));
            return nameIndex + StartColumnIndex;
        }

        private int GetRowIndex(string name)
        {
            int nameIndex;
            if (RowNames != null && RowNames.Contains(name))
                nameIndex = Array.IndexOf(RowNames,name);
            else
                throw VISite.Alerting.ThrowError("Can't Get Row: '" + name + "'. " + ((RowNames == null) ? "RowNames is Null" : ("Available RowNames: " + RowNames.Print(format:"'{0}'") + ")")));
            return nameIndex + StartRowIndex;
        }
        
        private string[] GetNumList(int count, int from = 1)
        {
            var result = new List<string>();
            for (int i = from; i < count + from; i++)
                result.Add(i.ToString());
            return result.ToArray();
        }

        public string Value { get {
            return "||X|" + ColumnNames.Print("|") + "||".LineBreak() + RowNames
                .Select(rowName => "||" + rowName + "||" + _allCells.Select(col => col.Value[rowName].Value).ToList().Print("|") + "||").ToList().Print("".LineBreak());
        } }

        public void SetValue<T1>(T1 value) { }


        private Cell<T> CreateCell(int colIndex, int rowIndex, string colName, string rowName)
        {
            return CreateCell(colIndex, rowIndex, Array.IndexOf(ColumnNames, colName) + 1, Array.IndexOf(RowNames, rowName) + 1, colName, rowName);
        }

        private Cell<T> CreateCell(int colIndex, int rowIndex, int colNum, int rowNum, string colName = "", string rowName = "")
        {
            var cell = CreateCell();

            if (_cellLocatorTemplate == null)
                _cellLocatorTemplate = (cell.HaveLocator())
                    ? cell.Locator
                    : By.XPath(".//tr[{1}]/td[{0}]");
            if (!cell.HaveLocator())
                cell.Locator = (_cellLocatorTemplate ?? By.XPath(".//tr[{1}]/td[{0}]")).FillByTemplate(colIndex, rowIndex);
            cell.InitSubElements();
            return new Cell<T>(cell.GetVIElement(), colNum, rowNum, colName, rowName);
        }

        public Table() { }
        public Table(string name = null, By tableLocator = null, By cellLocatorTemplate = null) : base(name)
        {
            Locator = tableLocator;
            _cellLocatorTemplate = cellLocatorTemplate;
        }

    }

}

File: Java/src/main/java/ru/viqa/ui_testing/elements/simpleElements/Table.java
package ru.viqa.ui_testing.elements.simpleElements;

import ru.viqa.ui_testing.common.funcInterfaces.*;
import ru.viqa.ui_testing.common.pairs.Pairs;
import ru.viqa.ui_testing.page_objects.PageObjectsInit;
import ru.viqa.ui_testing.elements.baseClasses.Table.Cell;
import ru.viqa.ui_testing.elements.baseClasses.Table.Columns;
import ru.viqa.ui_testing.elements.baseClasses.Table.Rows;
import ru.viqa.ui_testing.page_objects.VISite;
import ru.viqa.ui_testing.elements.baseClasses.*;
import ru.viqa.ui_testing.elements.interfaces.IHaveValue;
import ru.viqa.ui_testing.elements.interfaces.ITable;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.ArrayList;
import java.util.List;

import static java.util.Arrays.asList;
import static ru.viqa.ui_testing.common.utils.WebDriverByUtils.fillByTemplate;
import static ru.viqa.ui_testing.common.utils.PrintUtils.print;
import static ru.viqa.ui_testing.common.utils.LinqUtils.*;
import static ru.viqa.ui_testing.common.utils.StringUtils.*;

/**
 * Created by 12345 on 26.10.2014.
 */

public class Table<T extends IHaveValue> extends HaveValue implements ITable<T> {

    public Table() throws Exception{
        getColumns().table = this;
        getRows().table = this;
        //GetFooterFunc = t => t.FindElements(By.xpath("//tfoot/tr/td")).Select(el => el.Text).ToArray();
    }
    /*
    public Table(By tableLocator) {
        this();
        Locator = tableLocator;
    }

    public Table(String name = null, By tableLocator = null, By cellLocatorTemplate = null) {
        this()
        Name = name;
        Locator = tableLocator;
        _cellLocatorTemplate = cellLocatorTemplate;
    }*/
    private List<Cell<T>> _allCells = new ArrayList<>();
    public List<Cell<T>> getCells() throws Exception {
        for(String columnName : getColumns().headers())
            for(String rowName : getRows().headers())
                _allCells.add(cell(columnName, rowName));
        return _allCells;
    }

    private Columns<T> _columns = new Columns<T>();
    public Columns<T> getColumns() { return _columns; }
    public List<Cell<T>> getColumn(int colNum) throws Exception { return getRows().getColumn(colNum); }
    public List<Cell<T>> getColumn(String colName) throws Exception { return getRows().getColumn(colName); }
    public void setColumns(Columns<T> value) throws Exception { _columns.update(value); }
    public String[] getHeaders(FuncT<String[]> getHeadersAction) throws Exception {
        return doVIActionResult("Get Header", () -> getHeadersAction.invoke()); }

    private Rows<T> _rows = new Rows<T>();
    public Rows<T> getRows() { return _rows; }
    public List<Cell<T>> getRow(int rowNum) throws Exception { return getColumns().getRow(rowNum); }
    public List<Cell<T>> getRow(String rowName) throws Exception { return getColumns().getRow(rowName); }
    public void setRows(Rows<T> value) throws Exception { _rows.update(value); }

    public void setColumnHeaders(String[] value) { getColumns().setHeaders(value); }
    public void setRowHeaders(String[] value) { getRows().setHeaders(value); }
    public void setColCount(int value) { getColumns().setCount(value); }
    public void setRowCount(int value) { getRows().setCount(value); }

    protected String[] getFooterAction() throws Exception {
        return select(getWebElement().findElements(By.xpath("//tfoot/tr/td[1]")), WebElement::getText)
                .toArray(new String[1]);
    }
    protected String[] _footer;
    public void setFooter(String[] value) { _footer = value; }
    public String[] getFooter() throws Exception {
        if (_footer != null)
            return _footer;
        _footer = doVIActionResult("Get Footer", this::getFooterAction);
        if (_footer == null || _footer.length == 0)
            return null;
        getColumns().setCount(_footer.length);
        return _footer;
    }

    private Class<Text> clazz;
    public T getCellTemplate() throws Exception {
        return (T)(clazz.newInstance());
    }

    private By _cellLocatorTemplate;

    private T createCell() throws Exception {
        T cell = getCellTemplate();
        ((VIElement) cell).Context = new Pairs<>(ContextType.Locator, getLocator(), Context);
        return cell;
        }

    public Cell<T> cell(int colNum, int rowNum) throws Exception {
        int colIndex = colNum + getColumns().startIndex - 1;
        int rowIndex = rowNum + getRows().startIndex - 1;
        return addCell(colIndex, rowIndex, colNum, rowNum, "", "");
    }

    public Cell<T> cell(String colName, int rowNum) throws Exception {
        int colIndex = getColumnIndex(colName);
        int rowIndex = rowNum + getRows().startIndex - 1;
        return addCell(colIndex, rowIndex, asList(getColumns().headers()).indexOf(colName) + 1, rowNum, colName, "");
    }

    public Cell<T> cell(int colNum, String rowName) throws Exception {
        int colIndex = colNum + getColumns().startIndex - 1;
        int rowIndex = getRowIndex(rowName);
        return addCell(colIndex, rowIndex, colNum, asList(getRows().headers()).indexOf(rowName) + 1, "", rowName);
    }

    public Cell<T> cell(String colName, String rowName) throws Exception {
        int colIndex = getColumnIndex(colName);
        int rowIndex = getRowIndex(rowName);
        return addCell(colIndex, rowIndex, asList(getColumns().headers()).indexOf(colName) + 1,
                asList(getRows().headers()).indexOf(rowName) + 1, colName, rowName);
    }

    private Cell<T> addCell(int colIndex, int rowIndex, int colNum, int rowNum, String colName, String rowName) throws Exception {
        if (first(_allCells, cell -> cell.columnNum == colNum && cell.rowNum == rowNum) == null) {
            Cell<T> cell = createCell(colIndex, rowIndex, colNum, rowNum, colName, rowName);
            _allCells.add(cell);
            return cell;
        }
        return first(_allCells, cell -> cell.columnNum == colNum && cell.rowNum == rowNum).updateData(colName, rowName);
    }

    private List<Cell<T>> matches(List<Cell<T>> list, String pattern) throws Exception {
        return new ArrayList<>(where(list, cell -> ((IHaveValue) cell).getValue().matches(pattern)));
    }

    public List<Cell<T>> findCellsValues(String value) throws Exception {
        return new ArrayList<>(where(getCells(), cell -> ((IHaveValue) cell).getValue().equals(value)));
    }

    public List<Cell<T>> matchCellsValues(String pattern) throws Exception {
        return matches(getCells(), pattern);
    }

    public Cell<T> findFirstCellWithValue(String value) throws Exception {
        for (int colIndex = 1; colIndex <= getColumns().count(); colIndex++)
            for (int rowIndex = 1; rowIndex <= getRows().count(); rowIndex++) {
                Cell<T> cell = getCellFromValue(colIndex, rowIndex, value);
                if (cell != null)
                    return cell;
            }
        return null;
    }

    public Cell<T> findCellInColumn(int colIndex, String value) throws Exception {
        for (int rowIndex = 1; rowIndex <= getRows().count(); rowIndex++) {
            Cell<T> cell = getCellFromValue(colIndex, rowIndex, value);
            if (cell != null)
                return cell;
        }
        return null;
    }

    public Cell<T> findCellInColumn(String colName, String value) throws Exception {
        int colIndex = asList(getColumns().headers()).indexOf(colName) + 1;
        for (int rowIndex = 1; rowIndex <= getRows().count(); rowIndex++) {
            Cell<T> cell = getCellFromValue(colIndex, rowIndex, value);
            if (cell != null)
                return cell;
        }
        return null;
    }

    public List<Cell<T>> matchCellsInColumn(int colIndex, String pattern) throws Exception {
        return matches(getRow(colIndex), pattern);
    }

    public List<Cell<T>> matchCellsInColumn(String colname, String pattern) throws Exception{
        return matches(getRow(colname), pattern);
    }

    //Row filters
    public List<Cell<T>> matchCellsInRow(int rowIndex, String pattern) throws Exception {
        return matches(getColumn(rowIndex), pattern);
    }

    public List<Cell<T>> matchCellsInRow(String rowName, String pattern) throws Exception {
        return matches(getColumn(rowName), pattern);
    }

    private Cell<T> getCellFromValue(int colIndex, int rowIndex, String value) throws Exception {
        Cell<T> cell = cell(colIndex, rowIndex);
        return ((IHaveValue) cell).getValue().equals(value) ? cell : null;
    }

    public Cell<T> findCellInRow(int rowIndex, String value) throws Exception {
        for (int colIndex = 1; colIndex <= getColumns().count(); colIndex++) {
            Cell<T> cell = getCellFromValue(colIndex, rowIndex, value);
            if (cell != null)
                return cell;
        }
        return null;
    }

    public Cell<T> findCellInRow(String rowName, String value) throws Exception {
        int rowIndex = asList(getRows().headers()).indexOf(rowName) + 1;
        for (int colIndex = 1; colIndex <= getColumns().count(); colIndex++) {
            Cell<T> cell = getCellFromValue(colIndex, rowIndex, value);
            if (cell != null)
                return cell;
        }
        return null;
    }

    public List<Cell<T>> findColumnByRowValue(int rowIndex, String value) throws Exception {
        Cell<T> columnCell = findCellInRow(rowIndex, value);
        return columnCell != null ? getColumns().getRow(columnCell.columnNum) : null;
    }

    public List<Cell<T>> findColumnByRowValue(String rowName, String value) throws Exception {
        Cell<T> columnCell = findCellInRow(rowName, value);
        return columnCell != null ? getColumns().getRow(columnCell.columnNum) : null;
    }

    public List<Cell<T>> findRowByColumnValue(int colIndex, String value) throws Exception {
        Cell<T> rowCell = findCellInColumn(colIndex, value);
        return rowCell != null ? getRows().getColumn(rowCell.rowNum) : null;
    }

    public List<Cell<T>> findRowByColumnValue(String colName, String value) throws Exception {
        Cell<T> rowCell = findCellInColumn(colName, value);
        return rowCell != null ? getRows().getColumn(rowCell.rowNum) : null;
    }

    private int getColumnIndex(String name) throws Exception {
        int nameIndex;
        String[] headers = getColumns().headers();
        if (headers != null && asList(headers).contains(name))
            nameIndex = asList(headers).indexOf(name);
        else
            throw VISite.Alerting.throwError("Can't Get Column: '" + name + "'. " + ((headers == null)
                    ? "ColumnHeaders is Null" : ("Available ColumnHeaders: " + print(headers, ", ", "'{0}'") + ")")));
        return nameIndex + getColumns().startIndex;
    }

    private int getRowIndex(String name) throws Exception {
        int nameIndex;
        String[] headers = getRows().headers();
        if (headers != null && asList(headers).contains(name))
        nameIndex = asList(headers).indexOf(name);
        else
        throw VISite.Alerting.throwError("Can't Get Row: '" + name + "'. " +
                ((headers == null) ? "RowHeaders is Null" : ("Available RowHeaders: " + print(headers, ", ", "'{0}'") + ")")));
        return nameIndex + getRows().startIndex;
    }

    @Override
    public String getValueAction() throws Exception {
        return "||X|" + print(getColumns().headers(), "|") + "||" + LineBreak +
            print(new ArrayList(select(getRows().headers(),
                rowName -> "||" + rowName + "||" +
                    print(new ArrayList(select(where(getCells(),
                        cell -> cell.rowName.equals(rowName)),
                            cell -> ((IHaveValue)cell).getValue())), "|") + "||")), LineBreak);
    }

    @Override
    public void setValueAction(String string) {

    }

    private Cell<T> createCell(int colIndex, int rowIndex, String colName, String rowName) throws Exception {
        return createCell(colIndex, rowIndex, asList(getColumns().headers()).indexOf(colName) + 1,
            asList(getRows().headers()).indexOf(rowName) + 1, colName, rowName);
    }

    private Cell<T> createCell(int colIndex, int rowIndex, int colNum, int rowNum) throws Exception {
        return createCell(colIndex, rowIndex, colNum, rowNum, "", "");
    }
    private Cell<T> createCell(int colIndex, int rowIndex, int colNum, int rowNum, String colName, String rowName) throws Exception {
        VIElement cell = (VIElement)createCell();

        if (_cellLocatorTemplate == null)
            _cellLocatorTemplate = (cell.haveLocator())
                ? cell.getLocator()
                : By.xpath("//tr[{1}]/td[{0}]");
        if (!cell.haveLocator())
            cell.setLocator(fillByTemplate(_cellLocatorTemplate != null
                ? _cellLocatorTemplate : By.xpath("//tr[%s]/td[%s]"), rowIndex, colIndex));
        PageObjectsInit.initSubElements(cell);
        return new Cell((T)cell.getVIElement(), colNum, rowNum, colName, rowName);
    }
}
